### ISSUE TITLE
代码疑惑

### ISSUE THREAD

**Issue by qwpyyx** [shaoyijia/VAG#2]
Hi，作者您好。最近看了您的文章，感觉写的非常好，因此想要从代码上学习一下，感觉您的代码写的很清楚，于是仔细拜读了一下。
目前我看代码有三个比较疑惑地点，如果您可以解答的话感激不尽~
    我主要运行了原始版本，也就是finetune_classifier.py，使用的baseline是bart_classification_bart-base_vanilla。首先是finetune_classifier.py的taskcla设置成300的原因我不太清楚，此时num_labels=300，这样不是会让model的分类头变成(out_proj): Linear(in_features=768, out_features=300, bias=True)吗？我跑了fewrel数据集，他总共有80类，不应该设置成80吗？不然在train这部分，也就是my_bart_model.py的295行，logits = self.classification_head(sentence_representation)，logits的维度是{8，300}，其中8是batch_size的大小。我的理解这里不应该是{8，80}嘛？不然总感觉pred = outp.max(1)[1]这一步是不是也考虑到了其他的值呢？请问是我理解有问题吗？
![image](https://github.com/user-attachments/assets/e20bff4a-b7c0-41ee-8b1c-47e2a5fb2c58)
    第二个疑惑的点就是eval函数中的task_label_mask。我的理解他是对标签做掩码，防止有重复类出现。那是不是til_accuracy以及tid_pred_accuracy都跟CIL没关系，是TIL的情景嘞？我不是很明白til_accuracy和tid_pred_accuracy的具体含义及区别，能麻烦解惑一下嘛？
    第三个点就是第二个点的延申。在结果这一部分，我尝试理解了一下，但不知道是否正确，能否请您帮我判断一下，非常感谢。首先accumulated_acc_42代表随着任务的增加，模型在所有可见类上的整体性能。progressive_acc_42中对角线代表当前模型在当前任务的性能，从左到右则是当前模型在之前任务的test上的性能。如果是0，代表着当前任务训练完后的model对之前的任务完全遗忘。progressive_f1_42应该是同理。比较疑惑的是til和tid部分，比如tid_progressive_acc_42矩阵对角线全是1，不知道这代表什么（我目前只运行了4个任务），以及til_progressive中除了对角线，其他列也有值，说明没有完全遗忘，是因为这部分代码的功劳吗？
                if task_label_mask is not None:
                    masked_outp = outputs.logits * task_label_mask.to(outputs.logits.device)
                    til_pred = masked_outp.max(1)[1]
                    til_predictions = accelerator.gather(til_pred)
                    til_prediction_list += til_predictions.cpu().numpy().tolist()
![image](https://github.com/user-attachments/assets/874ea075-e5f7-4311-802e-5d3b66f1bb8e)

非常感谢您的回复！祝您科研顺利~

**Comment by shaoyijia** [comment]
> taskcla设置成300的原因我不太清楚

在class continual learning的设定里，未来会出现的类总数是不确定的（用固定数据集只是为了评测，如果在真实场景里，是不知道一共会有多少类的）。这里设成300这是一个较大的数，免去发现分类头不够需要扩增的麻烦。没有用到的分类头，不会被激活，所以不影响结果。实际中，如果在continual learning的过程中分类头不够了，可以增大这个值，用之前linear layer矩阵的参数初始化新linear layer的子矩阵（本质上linear layer就是个矩阵）.

> 不然总感觉pred = outp.max(1)[1]这一步是不是也考虑到了其他的值呢

确实考虑了，但是因为没有训练到的分类头输出会很小，不影响结果。

> 我的理解他是对标签做掩码，防止有重复类出现。那是不是til_accuracy以及tid_pred_accuracy都跟CIL没关系，是TIL的情景嘞？我不是很明白til_accuracy和tid_pred_accuracy的具体含义及区别，能麻烦解惑一下嘛？

> 比较疑惑的是til和tid部分，比如tid_progressive_acc_42矩阵对角线全是1，不知道这代表什么（我目前只运行了4个任务），以及til_progressive中除了对角线，其他列也有值，说明没有完全遗忘

是TIL的结果，这个主要是实验过程中方便来看错误到底是什么原因产生的。tid_pred_accuracy看的是当前类能不能至少被分到和他一起训的那组里，如果这个很高，但是主结果（progressive_acc, accumulated_acc)很低，那就说明是一组训练**内**的类的分界线模糊了；反之，则说明主要问题是一组组训练**间**的分界线模糊了。til_accuracy是假设我知道tid的结果，即消融掉了“一组组训练**间**的分界线模糊了”这个可能性的结果。
